Silicon/NXP/iMXPlatformPkg/Drivers/GopNullDxe/GopNullDxe.c

```diff
@@ -0,0 +1,271 @@
+  IN UINTN                              DestinationY,
+  IN UINTN                              Width,
+  IN UINTN                              Height,
+  IN UINTN                              Delta OPTIONAL
+  )
+{
+  UINT32  *FrameBuffer;
+  UINT32  BufferOffset;
+  UINT32  BufferWidth;
+  UINT32  FrameOffset;
+  UINT32  FrameWidth;
+  UINT32  i;
+
+  FrameBuffer = (UINT32 *)((UINTN)VidGopMode.FrameBufferBase);
+  FrameWidth = VidGopModeInfo.HorizontalResolution;
+  if (Delta == 0) {
+    BufferWidth = Width;
+  } else {
+    BufferWidth = Delta / PIXEL_BYTES;
+  }
+
+  if (BltOperation == EfiBltVideoFill) {
+    FrameOffset = FrameWidth * DestinationY + DestinationX;
+    for (i = DestinationY; i < DestinationY + Height; i++) {
+      SetMem32 (
+        FrameBuffer + FrameOffset,
+        Width * PIXEL_BYTES,
+        *(UINT32 *)BltBuffer
+      );
+      FrameOffset += FrameWidth;
+    }
+  } else if (BltOperation == EfiBltVideoToBltBuffer) {
+    FrameOffset = FrameWidth * SourceY + SourceX;
+    BufferOffset = BufferWidth * DestinationY + DestinationX;
+    for (i = SourceY; i < SourceY + Height; i++) {
+      CopyMem (
+        BltBuffer + BufferOffset,
+        FrameBuffer + FrameOffset,
+        Width * PIXEL_BYTES
+      );
+      FrameOffset += FrameWidth;
+      BufferOffset += BufferWidth;
+    }
+  } else if (BltOperation == EfiBltBufferToVideo) {
+    FrameOffset = FrameWidth * DestinationY + DestinationX;
+    BufferOffset = BufferWidth * SourceY + SourceX;
+    for (i = SourceY; i < SourceY + Height; i++) {
+      CopyMem (
+        FrameBuffer + FrameOffset,
+        BltBuffer + BufferOffset,
+        Width * PIXEL_BYTES
+      );
+      FrameOffset += FrameWidth;
+      BufferOffset += BufferWidth;
+    }
+  } else if (BltOperation == EfiBltVideoToVideo) {
+    FrameOffset = FrameWidth * DestinationY + DestinationX;
+    BufferOffset = FrameWidth * SourceY + SourceX;
+    for (i = SourceY; i < SourceY + Height; i++) {
+      CopyMem (
+        FrameBuffer + FrameOffset,
+        FrameBuffer + BufferOffset,
+        Width * PIXEL_BYTES
+      );
+      FrameOffset += FrameWidth;
+      BufferOffset += FrameWidth;
+    }
+  } else {
+    DEBUG ((DEBUG_ERROR, "%a: Not implemented %d\n",
+      __FUNCTION__, BltOperation));
+    return EFI_INVALID_PARAMETER;
+  }
+
+  return EFI_SUCCESS;
+}
+
+STATIC EFI_GRAPHICS_OUTPUT_PROTOCOL VidGop = {
+  VidGopQueryMode, // QueryMode
+  VidGopSetMode,   // SetMode
+  VidGopBlt,       // Blt
+  &VidGopMode    // Mode
+};
+
+EFI_STATUS
+GopNullDxeInitialize (
+  IN EFI_HANDLE         ImageHandle,
+  IN EFI_SYSTEM_TABLE   *SystemTable
+  )
+{
+  UINT32                  FrameBufferSize;
+  EFI_PHYSICAL_ADDRESS    FrameBufferBase;
+  EFI_STATUS              Status;
+
+  DEBUG ((DEBUG_INFO, "%a: Enter \n", __FUNCTION__));
+
+  // Allocate Frame Buffer
+  FrameBufferSize = mDisplayTiming.HActive * mDisplayTiming.VActive * mDisplayTiming.Bpp / 8;

```

Comment:
This allocates the buffer size based on the mDisplayTiming.Bpp value.  Everywhere else it assumes that Bpp is 32 and doesn't use mDisplayTiming.Bpp.  This should check for that and fail if mDisplayTiming.Bpp isn't 32?

Reply:
Good find. I think Bpp here should not be settable at all. Windows expects the pixels to be in BGRA8888 format. I'll switch Bpp to the PIXEL_BYTES define.